### ISSUE TITLE
ngtsc tries to compile non-exported components while ngc did not

### ISSUE THREAD

**Issue by devversion** [angular/angular#33724]
### Affected package

`@angular/compiler-cli` ngtsc.

### Problem description

Currently compiling a TS project which contains both test files and application code, NGC properly ignores non-exported components from test files.

When updating to v9, where ngtsc runs by default, the compilation will fail since `ngtsc` tries to compile the `@Component` annotations even for test components. Apparently since those test components are never part of a  `@NgModule`, the compilation will fail (e.g. component not found).

_accordion.spec.ts_
```ts
@Component({
  template: `<cdk-accordion>..</cdk-accordion>`
})
/** notice: not exported */ class TestComponent {}
```

One could argue that test files and application code shouldn't be part of the same compilation unit, but this has been working in View Engine and helped speeding up test compilation. There are more benefits like not having to use path mapping inside tests. i.e. relative imports to application code work just fine.

Another benefit is that the application/library components are compiled AOT. This is the primary reason why we follow this pattern (in `angular/components`).

**Note**: CLI projects follows the same pattern where tests and application code are part of the same compilation unit. The CLI just runs `tsc` instead of the Angular compiler (which we want for AOT)

### Possible solutions

* Saying that this is the expected behavior, though it will be a change in v9 that involves infrastructure changes. It doesn't seem easy to split into two compilation units, if tests and application code are located next to each other.
* Following NGC behavior by ignoring symbols which are not exported.
* Adding `jit: true` to every test component.

### Reproduction

https://github.com/devversion/ngtsc-test-component-analysis

**Comment by IgorMinar** [comment]
@devversion I'm a bit puzzled about why you are running ngc on the spec file. How does this speed up compilation?

**Comment by devversion** [comment]
@IgorMinar It speeds up compilation because we compile unit tests and application code within the same TypeScript compilation unit. If we'd split it into two compilation units, then we first build the application code, and have a second compilation unit for testing. The second compilation unit would start with a fresh `ts.Program` and would need to read parts of the previous compilation _again_.

It's generally common to build tests and application code in a same compilation unit. The CLI does the same by default, but just with the difference that it uses `tsc` while we want `ngc`. ngc will then process application code and AOT-compile it (e.g. generates JIT summaries), while the test code remains untouched since components are not exported (_=> and this is different in ngtsc_).

I realized that this is not quite what we do in `angular/components` since we do not actually use the JIT summaries, but the issue is still there for applications using AOT unit testing.

**Comment by devversion** [comment]
I think the frequency of this issue is _very_ low because AOT unit testing is not a common thing in Angular yet. Though it can be a common thing with Ivy since AOT unit testing will be very easy then.  

At that point, this issue will become relevant again and one of the possible solutions mentioned in this issue would need to be taken (unless we come up with something different).

**TL;DR**: It's a breaking change, but it's a rare one and there are two possible solutions. This issue will ultimately come up again in the future IMO when AOT unit testing with Ivy is a thing. I'm not sure if we should take any premature action yet. This feels like something we need to think more about (so that we can come up with a thought-through solution)
